src/Data/Graph/Haggle/Algorithms/DFS.hs

```diff
@@ -81,11 +81,9 @@ xdfsWith g nextVerts f roots
         True -> return acc
         False -> do
           setBitUnsafe bs (vertexId v)
-          nxt <- filterM (notVisited bs) (nextVerts v)
+          let nxt = nextVerts v
```

Comment:
Ah I see... it isn't exactly that the previous check handles it. It is that the recursive call will make the check anyway, so there is no savings gained with the `filterM`.